core/src/main/java/com/netflix/conductor/core/execution/WorkflowExecutor.java

```diff
@@ -78,121 +80,250 @@
-                logger.error("The input for the workflow {} cannot be NULL", workflowName);
-                throw new ApplicationException(INVALID_INPUT, "NULL input passed when starting workflow");
-            }
-
-            //Check if the workflow definition is valid
-            WorkflowDef workflowDefinition = metadataDAO.get(workflowName, workflowVersion);
-            if (workflowDefinition == null) {
-                logger.error("There is no workflow defined with name {} and version {}", workflowName, workflowVersion);
-                throw new ApplicationException(Code.NOT_FOUND, "No such workflow defined. name=" + workflowName + ", version=" + workflowVersion);
-            }
-
-            //because everything else is a system defined task
-            Set<String> missingTaskDefs = workflowDefinition.all().stream()
-                    .filter(task -> task.getType().equals(WorkflowTask.Type.SIMPLE.name()))
-                    .map(WorkflowTask::getName)
-                    .filter(task -> metadataDAO.getTaskDef(task) == null)
-                    .collect(Collectors.toSet());
-
-            if (!missingTaskDefs.isEmpty()) {
-                logger.error("Cannot find the task definitions for the following tasks used in workflow: {}", missingTaskDefs);
-                throw new ApplicationException(INVALID_INPUT, "Cannot find the task definitions for the following tasks used in workflow: " + missingTaskDefs);
-            }
-            //A random UUID is assigned to the work flow instance
-            String workflowId = IDGenerator.generate();
-
-            // Persist the Workflow
-            Workflow workflow = new Workflow();
-            workflow.setWorkflowId(workflowId);
-            workflow.setCorrelationId(correlationId);
-            workflow.setWorkflowType(workflowName);
-            workflow.setVersion(workflowVersion);
-            workflow.setInput(workflowInput);
-            workflow.setStatus(WorkflowStatus.RUNNING);
-            workflow.setParentWorkflowId(parentWorkflowId);
-            workflow.setParentWorkflowTaskId(parentWorkflowTaskId);
-            workflow.setOwnerApp(WorkflowContext.get().getClientApp());
-            workflow.setCreateTime(System.currentTimeMillis());
-            workflow.setUpdatedBy(null);
-            workflow.setUpdateTime(null);
-            workflow.setEvent(event);
-            workflow.setTaskToDomain(taskToDomain);
-            executionDAO.createWorkflow(workflow);
-            logger.info("A new instance of workflow {} created with workflow id {}", workflowName, workflowId);
-            //then decide to see if anything needs to be done as part of the workflow
-            decide(workflowId);
+        metadataMapperService.populateTaskDefinitions(workflowDefinition);
 
-            return workflowId;
+        // Obtain the names of the tasks with missing definitions
+        Set<String> missingTaskDefinitionNames = workflowDefinition.collectTasks().stream()
```

Comment:
Consider moving this logic to `metadataMapperService.populateTaskDefinitions`

Reply:
Moved and added test on top of that